File: NetworkServer/tests/WorldStateSnapshotCodecTests.cpp
#include "protocol/WorldStateSnapshotCodec.h"

#include <cstdint>
#include <exception>
#include <iostream>
#include <stdexcept>
#include <vector>

namespace
{
    void Require(
        const bool condition,
        const char* const message
    )
    {
        if (!condition)
        {
            throw std::runtime_error(message);
        }
    }

    template<typename Action>
    void RequireInvalidArgument(
        Action&& action,
        const char* const message
    )
    {
        try
        {
            action();
        }
        catch (const std::invalid_argument&)
        {
            return;
        }

        throw std::runtime_error(message);
    }

    void WorldStateSnapshotUsesStableWireLayout()
    {
        using namespace tdr::protocol;

        WorldEntityRecord record{};
        record.entityId = 0x01020304U;
        record.entityType = WorldEntityType::Player;
        record.lifecycle = WorldEntityLifecycle::Snapshot;
        record.flags = WorldEntityFlags::Active;
        record.positionX = 1.0F;
        record.positionY = -2.0F;
        record.rotationDegrees = 90.0F;
        record.currentHealth = 25U;
        record.maxHealth = 100U;
        record.bossPhase = 0U;

        WorldStateSnapshotPayload snapshot{};
        snapshot.entities.push_back(record);

        const auto encoded =
            WorldStateSnapshotCodec::Encode(snapshot);

        Require(
            encoded.size() ==
                kWorldStateSnapshotPrefixSize +
                kWorldEntityRecordSize,
            "World snapshot size is incorrect."
        );

        Require(
            encoded[0] == 0x00U &&
            encoded[1] == 0x00U &&
            encoded[2] == 0x00U &&
            encoded[3] == 0x01U,
            "World entity count is not network ordered."
        );

        Require(
            encoded[8] == 0x01U &&
            encoded[9] == 0x00U &&
            encoded[10] == 0x00U &&
            encoded[11] == 0x01U,
            "World entity identity fields are incorrect."
        );

        Require(
            encoded[24] == 0x00U &&
            encoded[25] == 0x19U &&
            encoded[26] == 0x00U &&
            encoded[27] == 0x64U,
            "World health fields are not network ordered."
        );
    }

    void WorldStateSnapshotRoundTripsMultipleEntities()
    {
        using namespace tdr::protocol;

        WorldStateSnapshotPayload snapshot{};

        WorldEntityRecord player{};
        player.entityId = 1U;
        player.entityType = WorldEntityType::Player;
        player.lifecycle = WorldEntityLifecycle::Snapshot;
        player.flags = WorldEntityFlags::Active;
        player.currentHealth = 10U;
        player.maxHealth = 10U;

        WorldEntityRecord boss{};
        boss.entityId = 2U;
        boss.entityType = WorldEntityType::Boss;
        boss.lifecycle = WorldEntityLifecycle::Snapshot;
        boss.flags = WorldEntityFlags::Active;
        boss.currentHealth = 0U;
        boss.maxHealth = 200U;
        boss.bossPhase = 2U;
        boss.flags = WorldEntityFlags::Dead;

        snapshot.entities = { boss, player };

        const auto encoded =
            WorldStateSnapshotCodec::Encode(snapshot);
        const auto decoded =
            WorldStateSnapshotCodec::Decode(
                encoded.data(),
                encoded.size());

        Require(
            decoded.entities.size() == 2U,
            "World snapshot entity count changed."
        );

        Require(
            decoded.entities[0].entityId == 1U &&
            decoded.entities[1].entityId == 2U,
            "World snapshot entities were not ordered by ID."
        );

        Require(
            decoded.entities[1].bossPhase == 2U &&
            decoded.entities[1].currentHealth == 0U &&
            decoded.entities[1].maxHealth == 200U,
            "Boss state did not round-trip."
        );
    }

    void WorldStateSnapshotRejectsInvalidHealth()
    {
        using namespace tdr::protocol;

        WorldEntityRecord record{};
        record.entityId = 1U;
        record.entityType = WorldEntityType::Enemy;
        record.lifecycle = WorldEntityLifecycle::Snapshot;
        record.flags = WorldEntityFlags::Active;
        record.currentHealth = 11U;
        record.maxHealth = 10U;
        record.enemyArchetype =
            NetworkEnemyArchetype::Basic;

        WorldStateSnapshotPayload snapshot{};
        snapshot.entities.push_back(record);

        RequireInvalidArgument(
            [&snapshot]()
            {
                static_cast<void>(
                    WorldStateSnapshotCodec::Encode(snapshot));
            },
            "Invalid world health was accepted."
        );
    }

    void WorldStateSnapshotPreservesEnemyArchetype()
    {
        using namespace tdr::protocol;

        WorldEntityRecord enemy{};
        enemy.entityId = 0x10000001U;
        enemy.entityType = WorldEntityType::Enemy;
        enemy.lifecycle = WorldEntityLifecycle::Snapshot;
        enemy.flags = WorldEntityFlags::Active;
        enemy.currentHealth = 1U;
        enemy.maxHealth = 1U;
        enemy.enemyArchetype =
            NetworkEnemyArchetype::Fast;

        WorldStateSnapshotPayload snapshot{};
        snapshot.entities.push_back(enemy);

        const auto encoded =
            WorldStateSnapshotCodec::Encode(snapshot);

        Require(
            encoded[
                kWorldStateSnapshotPrefixSize + 25U] ==
                static_cast<std::uint8_t>(
                    NetworkEnemyArchetype::Fast),
            "Enemy archetype is not at record offset 25."
        );

        for (std::size_t offset = 28U;
            offset < kWorldEntityRecordSize;
            ++offset)
        {
            Require(
                encoded[
                    kWorldStateSnapshotPrefixSize + offset] ==
                    0U,
                "World entity reserved byte is not zero."
            );
        }

        const auto decoded =
            WorldStateSnapshotCodec::Decode(
                encoded.data(),
                encoded.size());

        Require(
            decoded.entities[0].enemyArchetype ==
                NetworkEnemyArchetype::Fast,
            "Enemy archetype did not round-trip."
        );
    }

    void WorldStateSnapshotPreservesExperienceOrbAmount()
    {
        using namespace tdr::protocol;

        WorldEntityRecord orb{};
        orb.entityId = 0x40000001U;
        orb.entityType = WorldEntityType::ExperienceOrb;
        orb.lifecycle = WorldEntityLifecycle::Spawn;
        orb.flags = WorldEntityFlags::Active;
        orb.experienceAmount = 17U;

        WorldStateSnapshotPayload snapshot{};
        snapshot.entities.push_back(orb);

        const auto encoded =
            WorldStateSnapshotCodec::Encode(snapshot);
        const auto decoded =
            WorldStateSnapshotCodec::Decode(
                encoded.data(),
                encoded.size());

        Require(
            decoded.entities[0].experienceAmount == 17U,
            "Experience orb amount did not round-trip."
        );
    }

    void WorldStateSnapshotPreservesBossProjectileMetadata()
    {
        using namespace tdr::protocol;

        WorldEntityRecord projectile{};
        projectile.entityId = 0x30000001U;
        projectile.entityType = WorldEntityType::BossProjectile;
        projectile.lifecycle = WorldEntityLifecycle::Snapshot;
        projectile.flags = WorldEntityFlags::Active;
        projectile.directionX = 0.7071F;
        projectile.directionY = 0.7071F;
        projectile.projectileSpeed = 8.0F;
        projectile.projectileDamage = 12U;
        projectile.projectileSequence = 7U;

        WorldStateSnapshotPayload snapshot{};
        snapshot.entities.push_back(projectile);
        const auto encoded = WorldStateSnapshotCodec::Encode(snapshot);
        const auto decoded = WorldStateSnapshotCodec::Decode(
            encoded.data(),
            encoded.size());

        Require(decoded.entities[0].projectileDamage == 12U,
            "Boss projectile damage did not round-trip.");
        Require(decoded.entities[0].projectileSequence == 7U,
            "Boss projectile sequence did not round-trip.");
    }

    void WorldStateSnapshotValidatesBossPhase()
    {
        using namespace tdr::protocol;

        WorldEntityRecord boss{};
        boss.entityId = 7U;
        boss.entityType = WorldEntityType::Boss;
        boss.lifecycle = WorldEntityLifecycle::Snapshot;
        boss.flags = WorldEntityFlags::Active;
        boss.currentHealth = 10U;
        boss.maxHealth = 10U;

        WorldStateSnapshotPayload snapshot{};
        snapshot.entities.push_back(boss);

        for (const std::uint8_t phase : { 1U, 2U })
        {
            snapshot.entities[0].bossPhase = phase;
            const auto encoded = WorldStateSnapshotCodec::Encode(snapshot);
            const auto decoded = WorldStateSnapshotCodec::Decode(
                encoded.data(), encoded.size());
            Require(decoded.entities[0].bossPhase == phase,
                "Supported Boss phase did not round-trip.");
        }

        for (const std::uint8_t phase : { 0U, 3U })
        {
            snapshot.entities[0].bossPhase = phase;
            RequireInvalidArgument(
                [&snapshot]()
                {
                    static_cast<void>(
                        WorldStateSnapshotCodec::Encode(snapshot));
                },
                "Unsupported Boss phase was accepted.");
        }
    }

    void WorldStateSnapshotRejectsBossPhaseOnNonBoss()
    {
        using namespace tdr::protocol;

        WorldEntityRecord enemy{};
        enemy.entityId = 8U;
        enemy.entityType = WorldEntityType::Enemy;
        enemy.lifecycle = WorldEntityLifecycle::Snapshot;
        enemy.flags = WorldEntityFlags::Active;
        enemy.currentHealth = 10U;
        enemy.maxHealth = 10U;
        enemy.enemyArchetype = NetworkEnemyArchetype::Basic;
        enemy.bossPhase = 1U;

        WorldStateSnapshotPayload snapshot{};
        snapshot.entities.push_back(enemy);

        RequireInvalidArgument(
            [&snapshot]()
            {
                static_cast<void>(
                    WorldStateSnapshotCodec::Encode(snapshot));
            },
            "Non-Boss entity accepted a Boss phase.");
    }
}

int main()
{
    try
    {
        WorldStateSnapshotUsesStableWireLayout();
        WorldStateSnapshotRoundTripsMultipleEntities();
        WorldStateSnapshotRejectsInvalidHealth();
        WorldStateSnapshotPreservesEnemyArchetype();
        WorldStateSnapshotPreservesExperienceOrbAmount();
        WorldStateSnapshotPreservesBossProjectileMetadata();
        WorldStateSnapshotValidatesBossPhase();
        WorldStateSnapshotRejectsBossPhaseOnNonBoss();
        std::cout << "WorldStateSnapshotCodec tests passed.\n";
        return 0;
    }
    catch (const std::exception& exception)
    {
        std::cerr << exception.what() << '\n';
        return 1;
    }
}

File: Assets/Tests/EditMode/WorldStateSnapshotCodecTests.cs
using System;
using System.Collections.Generic;
using NUnit.Framework;
using TopDownRoguelike.Networking.Protocol;

namespace TopDownRoguelike.Tests.EditMode
{
    public sealed class WorldStateSnapshotCodecTests
    {
        [Test]
        public void EncodeThenDecode_PreservesStableWorldFields()
        {
            var player =
                new WorldEntityRecord(
                    1u,
                    NetworkEntityType.Player,
                    WorldEntityLifecycle.Snapshot,
                    WorldEntityFlags.Active,
                    1f,
                    -2f,
                    90f,
                    25,
                    100);

            var boss =
                new WorldEntityRecord(
                    2u,
                    NetworkEntityType.Boss,
                    WorldEntityLifecycle.Snapshot,
                    WorldEntityFlags.Dead,
                    5f,
                    6f,
                    180f,
                    0,
                    200,
                    2);

            var snapshot =
                new WorldStateSnapshotPayload(
                    new List<WorldEntityRecord>
                    {
                        boss,
                        player
                    });

            byte[] encoded =
                WorldStateSnapshotCodec.Encode(
                    snapshot);

            Assert.That(
                encoded.Length,
                Is.EqualTo(
                    WorldStateSnapshotCodec.PrefixSize +
                    2 * WorldStateSnapshotCodec.RecordSize));

            WorldStateSnapshotPayload decoded =
                WorldStateSnapshotCodec.Decode(
                    encoded);

            Assert.That(
                decoded.Entities,
                Has.Count.EqualTo(2));

            Assert.That(
                decoded.Entities[0].EntityId,
                Is.EqualTo(1u));

            Assert.That(
                decoded.Entities[1].EntityId,
                Is.EqualTo(2u));

            Assert.That(
                decoded.Entities[1].BossPhase,
                Is.EqualTo((byte)2));

            Assert.That(
                decoded.Entities[1].CurrentHealth,
                Is.EqualTo((ushort)0));

            Assert.That(
                decoded.Entities[1].IsDead,
                Is.True);
        }

        [Test]
        public void Encode_RejectsDeathFlagThatDoesNotMatchHealth()
        {
            var invalid =
                new WorldEntityRecord(
                    1u,
                    NetworkEntityType.Enemy,
                    WorldEntityLifecycle.Snapshot,
                    WorldEntityFlags.Dead,
                    0f,
                    0f,
                    0f,
                    10,
                    10,
                    0,
                    NetworkEnemyArchetype.Basic);

            var snapshot =
                new WorldStateSnapshotPayload(
                    new List<WorldEntityRecord>
                    {
                        invalid
                    });

            Assert.Throws<ArgumentException>(
                () => WorldStateSnapshotCodec.Encode(
                    snapshot));
        }

        [Test]
        public void Decode_RejectsTruncatedRecord()
        {
            byte[] truncated =
                new byte[WorldStateSnapshotCodec.PrefixSize];

            PacketCodec.WriteNetworkUInt32(
                truncated,
                0,
                1u);

            Assert.Throws<ArgumentException>(
                () => WorldStateSnapshotCodec.Decode(
                    truncated));
        }

        [Test]
        public void EncodeThenDecode_PreservesEnemyArchetypeAtOffset25()
        {
            var enemy =
                new WorldEntityRecord(
                    0x10000001u,
                    NetworkEntityType.Enemy,
                    WorldEntityLifecycle.Snapshot,
                    WorldEntityFlags.Active,
                    1f,
                    2f,
                    0f,
                    1,
                    1,
                    0,
                    NetworkEnemyArchetype.Fast);

            byte[] encoded =
                WorldStateSnapshotCodec.Encode(
                    new WorldStateSnapshotPayload(
                        new[]
                        {
                            enemy
                        }));

            int recordOffset =
                WorldStateSnapshotCodec.PrefixSize;

            Assert.That(
                encoded[recordOffset + 25],
                Is.EqualTo(
                    (byte)NetworkEnemyArchetype.Fast));

            for (int offset = 26;
                offset < WorldStateSnapshotCodec.RecordSize;
                offset++)
            {
                Assert.That(
                    encoded[recordOffset + offset],
                    Is.Zero);
            }

            WorldStateSnapshotPayload decoded =
                WorldStateSnapshotCodec.Decode(encoded);

            Assert.That(
                decoded.Entities[0].EnemyArchetype,
                Is.EqualTo(NetworkEnemyArchetype.Fast));
        }

        [Test]
        public void Encode_RejectsMissingOrMisplacedEnemyArchetype()
        {
            var enemyWithoutArchetype =
                new WorldEntityRecord(
                    0x10000001u,
                    NetworkEntityType.Enemy,
                    WorldEntityLifecycle.Snapshot,
                    WorldEntityFlags.Active,
                    0f,
                    0f,
                    0f,
                    3,
                    3);

            var playerWithArchetype =
                new WorldEntityRecord(
                    1u,
                    NetworkEntityType.Player,
                    WorldEntityLifecycle.Snapshot,
                    WorldEntityFlags.Active,
                    0f,
                    0f,
                    0f,
                    5,
                    5,
                    0,
                    NetworkEnemyArchetype.Basic);

            Assert.Throws<ArgumentException>(
                () => WorldStateSnapshotCodec.Encode(
                    new WorldStateSnapshotPayload(
                        new[]
                        {
                            enemyWithoutArchetype
                        })));

            Assert.Throws<ArgumentException>(
                () => WorldStateSnapshotCodec.Encode(
                    new WorldStateSnapshotPayload(
                        new[]
                        {
                            playerWithArchetype
                        })));
        }
    }
}
